### ISSUE TITLE
Non-symmetric solution (but it should be)

### ISSUE THREAD

**Issue by tgregov** [tgregov/Multiphysics#28]
https://github.com/tgregov/Multiphysics/blob/ac52b6d46c05f6bed60596e068151b3769a9836d/srcs/buildFlux.cpp#L154
by putting in the BC aera `gx[j]  = 1` and `gy[j]=0`, we get a non-symmetric results => it may come from those two lines.
In particular, we get `gy[j] != 0`

**Comment by tgregov** [comment]
Edit: it might also come from the matrices product in the F computation 😞

**Comment by tgregov** [comment]
```C
unsigned int frontJ = std::get<1>(edge.edgeInFront);
```
I don't think we have thought about the orientation of the edge (but maybe I'm wrong). But since by looking manually the orientation is inverted, shouldn't we put 
```C
unsigned int frontJ = std::get<2>(edge.edgeInFront);
```
instead ? @ImperatorS79 what do you think ?

**Comment by ImperatorS79** [comment]
That will return a boolean... Remember j is the index of the edge in the element (0, 1 or 2).

**Comment by tgregov** [comment]
Still, there is something wrong here, in the sense that frontJ is the same for all j

**Comment by ImperatorS79** [comment]
I think that the chosen name is confusing.
The `edgeInFront` tuple simply gives : 0: the index of the element in which the edge in front is, 1: the index of that edge inside that element (0, 1, 2), 2: a boolean to see if the nodes tags are in the same order in both edge.

We called the variable `frontJ` because we thought it would work, but it has indeed nothing to do with the j in that function.

**Comment by tgregov** [comment]
I have tried something like this: (sorry for the indentation, copy-paste it somewhere...)
```C
				// loop over the SF
				for(unsigned int j = 0 ; j < elmProp2D.nSF ; ++j)
				{

					//[TO DO] VERY TEMPORARY SOLUTION !!!!! ONLY WORKS FOR T3
                    // the side s contains the SF j, j+1
                    if(j == s || (s+1)%3 == (j)%3){
                    	unsigned int frontOffsetInU = entity.elements[std::get<0>(edge.edgeInFront)].offsetInU;
                    	unsigned int frontJ;
                    	if(j == s)
                    	{
                    		// it is inverted
							frontJ = (std::get<1>(edge.edgeInFront) + 1) % 3;
                    	}
                    	else // j == s+1
                    	{
                    		// it is inverted
                    		frontJ = std::get<1>(edge.edgeInFront);
                    	}

						if (std::get<0>(edge.edgeInFront) == -1)
						{
							if(elm == 1)
							{
								gx[j] = 1.0;
								gy[j] = 0.0;
							}
							else
							{
								gx[j] = 1.0;
								gy[j] = 0.0;
							}
							/*
							if(edge.nodeTags.first == )
								gx[j] = -(factor*fx[element.offsetInU + j] + fx[frontOffsetInU + frontJ])/2
								- C*element.edges[s].normal.first*(u[element.offsetInU + j] - valueAtBC())/2;

								gy[j] = -(factor*fy[element.offsetInU + j] + fy[frontOffsetInU + frontJ])/2
								- C*element.edges[s].normal.second*(u[element.offsetInU + j] - u[frontOffsetInU + frontJ])/2;
								*/
						}
						else
						{
							// BUg
							gx[j] = -(factor*fx[element.offsetInU + j] + fx[frontOffsetInU + frontJ])/2
									- C*element.edges[s].normal.first*(u[element.offsetInU + j] - u[frontOffsetInU + frontJ])/2;

							gy[j] = -(factor*fy[element.offsetInU + j] + fy[frontOffsetInU + frontJ])/2
									- C*element.edges[s].normal.second*(u[element.offsetInU + j] - u[frontOffsetInU + frontJ])/2;
							//std::cout << "gy" << gy[j] << std::endl;
							//gx[j]=0.1;
							//gy[j]=0;
						}
                    }
                    else
                    {
                    	gx[j] = 0.0;
						gy[j] = 0.0;
                    }
```

**Comment by tgregov** [comment]
It "seems" to work, but makes a LOT of assumptions (T3 etc). A more elegant solution would be to store the edge in front (the real one, let's say the jInFront) in Mesh2D...

**Comment by ImperatorS79** [comment]
What do you want to store exactly that is not currently stored ?

**Comment by tgregov** [comment]
The index j (in our case 0, 1, 2) of the element "in front". Using that, we can just type the offset of the element + this value when considering the u of this neighbour

**Comment by tgregov** [comment]
in fact, this corresponds to the frontJ, but the real frontJ value that we were expecting

**Comment by ImperatorS79** [comment]
The index 0, 1, 2 is already stored.

**Comment by tgregov** [comment]
It is the index of the edge, not of the node

**Comment by tgregov** [comment]
For T3, the conversion (index of the edge)+(inverted) -> (node) is what I have done in the code I've sent you, but it is not general at all + might become more complicated to determine later

**Comment by ImperatorS79** [comment]
Yes, but you can easily determine the index of the nodes, which is 0 or 1, which is the same than in `edge.nodeTags` if `inverted == false`, and inverted otherwise

**Comment by tgregov** [comment]
so you want to access the fields using something like frontOffsetInU+frontJ+(inverted == false) ? But you still don't consider that this should be different for different j values. Maybe I don't really understand how you want to do that.

**Comment by tgregov** [comment]
And btw there is a mistake in the code I've sent you. Here is what I wanted to do:
(i) check that the j belongs to a current edge => does s==j or s+1==j ? [modulo the # of nodes]
(ii) using the order (at the left/at the right), depending on the two cases, we can determine the frontJ value I'm talking about

**Comment by ImperatorS79** [comment]
I think we are not understanding each other.

In an `Edge` struct, you have a tuple indicating the index of the other element and the index of the edge inside that element, in which you will find the same edge. the index of the edge in the other element can go form 0 to the number of edge inside that element.

Then, by edge, you have two nodes (linear element), and the boolean is indcating if the order of nodes in `nodesTags` is the same as in the other edge `nodesTags`.

So, if what you want is the position of that nodes in the u vector, you the the three info's.

**Comment by tgregov** [comment]
The problem that I have is that you still need to make (small) computations to know which node you are considering. Meaning that saying that the orientation is inverted is not enough: you also need to compute the link indexJ -> position in edge (first or second), as I've said in my previous message. This computation should be done in Mesh2D for 2 reasons: it is useless to do that everythime + to generalize it to non T3 elements, it is better that it is not in builFlux

**Comment by ImperatorS79** [comment]
Definitely yes.

**Comment by tgregov** [comment]
So we have stored all the necessary information, but maybe there is a smarter way to store it, or other values that are more interesting to store

**Comment by ImperatorS79** [comment]
Let's just code it and see if it works first. We will certainly either see a smartest way while coding or later if any exists.

**Comment by tgregov** [comment]
You can just copy-paste the previous code I've put here, and check if it makes sense 😄

**Comment by ImperatorS79** [comment]
I meant the storage in Mesh2D. But yes I will check your code + the modification you put in comment. Does it gives a symmetric solution ?

**Comment by tgregov** [comment]
I understood in your previous message that you didn't want to change the structure in Mesh2D, but it is obviously better to change it there ! And my code gives a symmetric solution.

**Comment by ImperatorS79** [comment]
No, I am perfectly open to store things in Mesh2D if it allows us to not do computation later on. I just wanted to be sure of what we were talking about ^^.

**Comment by tgregov** [comment]
Ok is this OK if you code this part (since you're more familiar with the Mesh2D.hpp structure as we have seen before 😄 ) ? Then I can review it. In the meantime I will check with the other groups/manually the M and S matrices

**Comment by ImperatorS79** [comment]
To be sure, we want to store, for each nodeInFront, a value `j + (inverted == true)` ?

**Comment by tgregov** [comment]
We want that for each node in each edge, we can get its index in the adjacent element. But maybe there are others way to do it, so do not hesitate to store it as you wish, as soon as it easy to use in the code

**Comment by tgregov** [comment]
The matrix [M] is ok. I have spotted a mistake in [S]; I'm on it 🤓

**Comment by ImperatorS79** [comment]
Ok I have modified Mesh2D, now there is a `nodeIndexEdgeInFront` to store the index of the nodes in the edge in front (for T3 is is 0-1 or 1-0). I have then deleted the `std::tuple` and replaced it with an `std::pair`. I have also replaced the std::pair for the `nodeTags `and `nodeCoordinate` by a vector, so that in the future we can more easily modify the code for non-linear edge. There is a magic number in `addEntity `where I added a `[TO DO]` which is directly link to that issue.

@tgregov Can I let you the modification of `buildFlux.cpp` ? I am not really familiar enough with the flux. However I can optimize it after (for the comparison weak/strong that are done at each step, I have already some commits locally ^^).
